pkg/synchronization-controller/synchronization-controller.go

```diff
@@ -0,0 +1,795 @@
+			return err
+		}
+	}
+	if outboundConnection == nil {
+		log.Log.Object(vmi).V(4).Info("no synchronization connection found for source, doing nothing")
+		return nil
+	}
+	vmiStatusJson, err := json.Marshal(vmi.Status)
+	if err != nil {
+		return err
+	}
+	client := syncv1.NewSynchronizeClient(outboundConnection.grpcClientConnection)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(s.timeout)*time.Second)
+	defer cancel()
+
+	if _, err := client.SyncSourceMigrationStatus(ctx, &syncv1.VMIStatusRequest{
+		MigrationID: outboundConnection.migrationID,
+		VmiStatus: &syncv1.VMIStatus{
+			VmiStatusJson: vmiStatusJson,
+		},
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *SynchronizationController) handleTargetState(vmi *virtv1.VirtualMachineInstance) error {
+	if vmi.Status.MigrationState == nil {
+		// No migration state, don't do anything
+		return nil
+	}
+	if vmi.Status.MigrationState.TargetState == nil || vmi.Status.MigrationState.SourceState == nil {
+		// No migration state, don't do anything
+		return nil
+	}
+
+	var outboundConnection *SynchronizationConnection
+	var err error
+	sourceState := vmi.Status.MigrationState.SourceState
+	targetState := vmi.Status.MigrationState.TargetState
+	if targetState.SyncAddress == nil || *targetState.SyncAddress == "" {
+		syncAddress, err := s.getMyURL()
+		if err != nil {
+			return err
+		}
+		targetState.SyncAddress = &syncAddress
+	}
+
+	if sourceState.SyncAddress != nil && targetState.MigrationUID != "" {
+		if outboundConnection, err = s.getOutboundTargetConnection(vmi, vmi.Status.MigrationState); err != nil {
+			return err
+		}
+	}
+	if outboundConnection == nil {
+		log.Log.Object(vmi).V(4).Info("no synchronization connection found for target, doing nothing")
+		return nil
+	}
+	vmiStatusJson, err := json.Marshal(vmi.Status)
+	if err != nil {
+		return err
+	}
+	client := syncv1.NewSynchronizeClient(outboundConnection.grpcClientConnection)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(s.timeout)*time.Second)
+	defer cancel()
+
+	_, err = client.SyncTargetMigrationStatus(ctx, &syncv1.VMIStatusRequest{
+		MigrationID: outboundConnection.migrationID,
+		VmiStatus: &syncv1.VMIStatus{
+			VmiStatusJson: vmiStatusJson,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *SynchronizationController) getMigrationForVMI(vmi *virtv1.VirtualMachineInstance) (*virtv1.VirtualMachineInstanceMigration, error) {
+	objects, err := s.migrationInformer.GetIndexer().ByIndex("byVMIName", vmi.Name)
+	if err != nil {
+		return nil, err
+	}
+	if len(objects) > 0 {
+		count := 0
+		var res *virtv1.VirtualMachineInstanceMigration
+		for _, migrationObj := range objects {
+			migration, ok := migrationObj.(*virtv1.VirtualMachineInstanceMigration)
+			if !ok {
+				return nil, fmt.Errorf("not a virtual machine instance migration")
+			}
```

Comment:
We usually assume successful casts from informer objects, but I guess it can't hurt to double-check :)